gui/job-details.lua

```diff
@@ -0,0 +1,635 @@
+        self:setMaterial(obj, -1, -1)
+    end
+end
+
+function JobDetails:onChangeIType()
+    local idx, obj = self.list:getSelected()
+    guimat.ItemTypeDialog{
+        prompt = 'Please select a new item type for input '..idx,
+        none_caption = 'any item',
+        item_filter = curry(dfhack.job.isSuitableItem, obj.iobj),
+        on_select = self:callback('setItemType', obj)
+    }:show()
+end
+
+function JobDetails:canChangeMat()
+    local idx, obj = self.list:getSelected()
+    return obj ~= nil and not is_caste_mat(obj.iobj)
+end
+
+function JobDetails:setMaterial(obj, mat_type, mat_index)
+    if  obj.index == 0
+    and self.job.mat_type == obj.iobj.mat_type
+    and self.job.mat_index == obj.iobj.mat_index
+    and self.job.job_type ~= df.job_type.PrepareMeal
+    then
+        self.job.mat_type = mat_type
+        self.job.mat_index = mat_index
+    end
+
+    obj.iobj.mat_type = mat_type
+    obj.iobj.mat_index = mat_index
+end
+
+function JobDetails:findUnambiguousItem(iobj)
+    local count = 0
+    local itype
+
+    for i = 0,df.item_type._last_item do
+        if dfhack.job.isSuitableItem(iobj, i, -1) then
+            count = count + 1
+            if count > 1 then return nil end
+            itype = i
+        end
+    end
+
+    return itype
+end
+
+function JobDetails:onChangeMat()
+    local idx, obj = self.list:getSelected()
+
+    if obj.iobj.item_type == -1 and obj.iobj.mat_type == -1 then
+        -- If the job allows only one specific item type, use it
+        local vitype = self:findUnambiguousItem(obj.iobj)
+
+        if vitype then
+            obj.iobj.item_type = vitype
+        else
+            dlg.showMessage(
+                'Bug Alert',
+                { 'Please set a specific item type first.\n\n',
+                  'Otherwise the material will be matched\n',
+                  'incorrectly due to a limitation in DF code.' },
+                COLOR_YELLOW
+            )
+            return
+        end
+    end
+
+    guimat.MaterialDialog{
+        prompt = 'Please select a new material for input '..idx,
+        none_caption = 'any material',
+        mat_filter = function(mat,parent,mat_type,mat_index)
+            return dfhack.job.isSuitableMaterial(obj.iobj, mat_type, mat_index, obj.iobj.item_type)
+        end,
+        on_select = self:callback('setMaterial', obj)
+    }:show()
+end
+
+function JobDetails:canChangeTrait()
+    local idx, obj = self.list:getSelected()
+    return obj ~= nil and not is_caste_mat(obj.iobj)
+end
+
+function JobDetails:onChangeTrait()
+    local idx, obj = self.list:getSelected()
+    guimat.ItemTraitsDialog{
+        job_item = obj.iobj,
+        prompt = 'Please select traits for input '..idx,
+        none_caption = 'no traits',
+    }:show()
+end
+
+function JobDetails:onResetChanges()
+    for _, obj in pairs(self.list.choices) do
+        local stored_obj = self.stored[obj.index]
+
+        local item_type = stored_obj.item_type
```

Comment:
Attempting to reset changes for me results in this error:
```
gui/job-details.lua:419: attempt to index a nil value (local 'stored_obj')
```

Reply:
looking through the code, it looks like storeInitialProperties is checking `job.items` instead of `job.job_items`

Reply:
the `job.job_items` vs `job.items` is the bane of my existence 😣

because there are two structures involved, manager order and job, and one of them has both but uses the one that's distinct.

I think I fixed it now.